lib/ukfalloc: Export stats via uk_store

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]


### Additional configuration

Introduce `CONFIG_LIBUKFALLOC_STATS` to enable exporting stats via `ukstore`.

### Description of changes


This series adds `ukfalloc` the capability to export aggregate and per-allocator stats via uk_store. The stats are exported by the implementation, in this case `ukfallocbuddy`.

GitHub-Depends-On: #939
GitHub-Depends-On: #938

michpappas commented:
@marcrittinghaus we should consider whether it would make more sense for `uk_falloc` to track pages rather than bytes.

craciunoiuc commented on lib/ukfalloc/stats.c:
```diff
@@ -0,0 +1,116 @@
+static const struct uk_store_entry *dyn_entries[] = {
+	UK_STORE_ENTRY(UK_FALLOC_STATS_FREE_MEMORY, "free_memory", u64,
+		       get_free_memory, NULL, NULL),
+	UK_STORE_ENTRY(UK_FALLOC_STATS_TOTAL_MEMORY, "free_memory", u64,
```
shouldn't this be `"total_memory"`?

michpappas replied:
Fixed, thanks

craciunoiuc commented on lib/ukfallocbuddy/fallocbuddy.c:
```diff
@@ -1451,6 +1465,10 @@ static int bfa_do_addmem(struct buddy_framealloc *bfa, void *metadata,
 	bfa_zone_add(bfa, zone);
 	bfa->fa.total_memory += len;
 
+#ifdef CONFIG_LIBUKFALLOC_STATS
+	uk_falloc_stats_global.total_memory += len;
+#endif /* CONFIG_LIBUKFALLOC_STATS */
```
@michpappas are you sure this is okay? Total memory never decreases (like `free_memory`)

michpappas replied:
I think so. The API provides an `addmem` operation but no ability to remove memory. The `alloc` and `free` functions operate on previously added memory.

craciunoiuc reviewed (approved):
All good here. Masan!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

michpappas committed:
lib/ukfalloc: Export allocator stats

Export system-wide and per-allocator statistics via libukstore. This
is controlled by a hidden Kconfig option selected by frame allocator
libraries that wish to export metrics.

Entry IDs of the exported metrics are provided in include/uk/falloc_store.h

Since the frame allocator is initialized before the heap, early calls
to uk_falloc_init_stats() are deferred to init. This is not visible to
the caller.

Provide uk_falloc_init() helper that also initializes stats.

Checkpatch-Ignore: MACRO_ARG_REUSE
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukfallocbuddy: Populate ukfalloc stats

Register into ukfalloc via helper. Update system-wide
ukralloc stats.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
> @marcrittinghaus we should consider whether it would make more sense for `uk_falloc` to track pages rather than bytes.

To answer my own question, that's probably not a good idea given that some architectures like Arm can have a variable translation granule.

michpappas commented:
Rebased from `staging`

adinamariav reviewed (approved):
Reviewed-by: Adina-Maria Vaman adinamariav22@gmail.com

michpappas commented:
## Test Code

This test creates a second frame allocator from which it periodically allocates / frees pages to demonstrate the variation in exported stats.

```c
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>

#include <uk/event.h>
#include <uk/falloc/store.h>
#include <uk/fallocbuddy.h>
#include <uk/libid.h>
#include <uk/list.h>
#include <uk/store.h>

#define DIRECTMAP_AREA_START	0xffffff8000000000 /* -512 GiB */
#define DIRECTMAP_AREA_END	0xffffffffffffffff
#define DIRECTMAP_AREA_SIZE	(DIRECTMAP_AREA_END - DIRECTMAP_AREA_START + 1)

/* plat/kvm/x86_64 bpt map up to 4GiB. Configure
 * QEMU to use 5GiB and configure the 2nd falloc
 * instnance to start at 4GiB.
 */
#define FA2_BASE		(4 * 1024UL * 1024UL * 1024UL)
#define FA2_LEN			(1024UL * 1024UL * 1024UL)

struct uk_falloc_test_obj {
	struct uk_store_object *obj;
	struct uk_list_head list_head;
};

/* List of test objects */
static UK_LIST_HEAD(tobj_list); 

static const struct uk_store_entry *e_sys_free;
static const struct uk_store_entry *e_sys_total;

static struct uk_falloc *fa2;
static __paddr_t fa2_alloc_start;

static unsigned int test_count;

static inline __vaddr_t
x86_directmap_paddr_to_vaddr(__paddr_t paddr)
{
	UK_ASSERT(paddr < DIRECTMAP_AREA_SIZE);
	return (__vaddr_t)paddr + DIRECTMAP_AREA_START;
}

static inline __paddr_t
x86_directmap_vaddr_to_paddr(__vaddr_t vaddr)
{
	UK_ASSERT(vaddr >= DIRECTMAP_AREA_START);
	UK_ASSERT(vaddr < DIRECTMAP_AREA_END);
	return (__paddr_t)(vaddr - DIRECTMAP_AREA_START);
}

static inline int
falloc_add_mem(struct uk_falloc *fa, __paddr_t start, __sz len)
{
	unsigned long pages = len >> PAGE_SHIFT;
	void *fa_meta;
	__sz fa_meta_size;
	__vaddr_t dm_off;
	int rc;

	UK_ASSERT(start <= __PADDR_MAX - len);
	UK_ASSERT(ukarch_paddr_range_isvalid(start, len));

	fa_meta = (void *)x86_directmap_paddr_to_vaddr(start);
	fa_meta_size = uk_fallocbuddy_metadata_size(pages);

	if (unlikely(fa_meta_size >= len))
		return -ENOMEM;

	start = PAGE_ALIGN_UP(start + fa_meta_size);
	pages = (len - fa_meta_size) >> PAGE_SHIFT;

	dm_off = x86_directmap_paddr_to_vaddr(start);

	UK_ASSERT(fa->addmem);

	rc = fa->addmem(fa, fa_meta, start, pages, dm_off);
	if (unlikely(rc))
		return rc;

	fa2_alloc_start = start;

	return 0;
}

int ukstore_falloc_test_init(void)
{
	__paddr_t start;
	__sz fa_size;
	__sz len;
	int rc;

	/* Get static entries */
	e_sys_free  = uk_store_static_entry_get(__UKLIBID_libukfalloc,
						UK_FALLOC_STATS_FREE_MEMORY);
	e_sys_total = uk_store_static_entry_get(__UKLIBID_libukfalloc,
						UK_FALLOC_STATS_TOTAL_MEMORY);

	/* Create a second frame allocator to demonstrate
	 * varible usage.
	 */
	start = FA2_BASE;
	len = FA2_LEN;

	fa2 = (struct uk_falloc *)x86_directmap_paddr_to_vaddr(start);

	fa_size = ALIGN_UP(uk_fallocbuddy_size(), 8);
	if (unlikely(fa_size >= len))
		return -ENOMEM;

	rc = uk_fallocbuddy_init(fa2);
	if (unlikely(rc))
		return rc;

	rc = falloc_add_mem(fa2, start + fa_size, len - fa_size);
	if (unlikely(rc))
		return rc;

	return 0;
}

/* Allocate / free frames in 4 + 4 iterations */
void fa2_alloc_free(void)
{
	unsigned long frames;
	unsigned long iter;
	__paddr_t base;

	frames = ((FA2_BASE + FA2_LEN - fa2_alloc_start) >> PAGE_SHIFT) / 4;
	iter = test_count++ % 8;
	base = fa2_alloc_start + (iter % 4) * (frames << PAGE_SHIFT);

	if (iter < 4)
		fa2->falloc(fa2, &base, frames, 0);
	else if (iter >= 4)
		fa2->ffree(fa2, base, frames);
}

void ukstore_falloc_test(void)
{
	const struct uk_store_entry *e_free;
	const struct uk_store_entry *e_total;
	struct uk_falloc_test_obj *tobj;
	__u64 free;
	__u64 total;

	/* Print static entries */
	uk_store_get_value(e_sys_free,  u64, &free);
	uk_store_get_value(e_sys_total, u64, &total);

	printf("sys:   : 0x0%012lx / 0x%012lx (%ld%%)\n",
	       free, total,
	       (free * 100) / total);

	/* Allocate / free fa2 frames once in a while to demonstrate
	 * variation in usage
	 */
	fa2_alloc_free();

	uk_list_for_each_entry(tobj, &tobj_list, list_head) {
		e_free  = uk_store_obj_entry_get(tobj->obj,
						 UK_FALLOC_STATS_FREE_MEMORY);
		e_total = uk_store_obj_entry_get(tobj->obj,
						 UK_FALLOC_STATS_TOTAL_MEMORY);

		uk_store_get_value(e_free, u64, &free);
		uk_store_get_value(e_total, u64, &total);

		printf("%s: 0x0%012lx / 0x%012lx (%ld%%)\n",
			tobj->obj->name,
			free, total,
			(free * 100) / total);
	}
	printf("\n");
}

static int handle_create_object(void *arg)
{
	struct uk_store_event_data *data = (struct uk_store_event_data *)arg;
	struct uk_falloc_test_obj *tobj;
	struct uk_store_object *obj;

	/* Is this coming from libukfalloc? */
	if (data->library_id != __UKLIBID_libukfalloc)
		return UK_EVENT_NOT_HANDLED; /* continue to next handler */

	/* Increment the refcount */
	obj = uk_store_obj_acquire(data->library_id, data->object_id);
	if (unlikely(!obj))
		return UK_EVENT_HANDLED; /* The object was already released */

	/* Add to list of test objects */
	tobj = malloc(sizeof(*tobj));
	if (unlikely(!tobj))
		UK_CRASH("Out of memory\n");

	tobj->obj = obj;
	uk_list_add_tail(&tobj->list_head, &tobj_list);

	return UK_EVENT_HANDLED;
}

UK_EVENT_HANDLER(UKSTORE_EVENT_CREATE_OBJECT, handle_create_object);
```

## Integration with `app-helloworld`:

Update `main()` to call `ukstore_falloc_test_init()` once, and `ukstore_falloc_test()` periodically, similarly to:

```diff
int main(int argc, char *argv[])
{
#if CONFIG_APPHELLOWORLD_PRINTARGS || CONFIG_APPHELLOWORLD_SPINNER
	int i;
#endif

+ #if CONFIG_LIBUKFALLOC_STATS
+ 	ukstore_falloc_test_init();
+ #endif /* CONFIG_LIBUKFALLOC_STATS */

	printf("Hello world!\n");

#if CONFIG_APPHELLOWORLD_PRINTARGS
	printf("Arguments: ");
	for (i=0; i<argc; ++i)
		printf(" \"%s\"", argv[i]);
	printf("\n");
#endif /* CONFIG_APPHELLOWORLD_PRINTARGS */

#if CONFIG_APPHELLOWORLD_SPINNER
	i = 0;
	printf("\n\n\n");
	for (;;) {
+ #if CONFIG_LIBUKFALLOC_STATS
+ 	ukstore_falloc_test();
+ #else /* CONFIG_LIBUKFALLOC_STATS */
		i %= (monkey3_frame_count * 3);
		printf("\r\033[2A %s \n", monkey3[i++]);
		printf(" %s \n",          monkey3[i++]);
		printf(" %s ",            monkey3[i++]);
+ #endif /* CONFIG_LIBUKFALLOC_STATS */
		fflush(stdout);
		millisleep(250);
	}
#endif /* CONFIG_APPHELLOWORLD_SPINNER */

	return 0;
}
```

## QEMU Configuration

This creates a second frame allocator at the 4GiB map boundary and assigns it 1GiB, so it requires configuring QEMU to use 5GiB.

michpappas commented:
@skuenzer this is ready for review

skuenzer commented on lib/ukfalloc/include/uk/falloc/store.h:
```diff
@@ -0,0 +1,35 @@
+/* Initialize stats for a new instance of uk_falloc
+ *
+ * This function should be called when instantiating
+ * a new allocator.Early calls to this function made
```
Very minor: Just a space missing.

skuenzer commented on lib/ukfalloc/Config.uk:
```diff
@@ -5,4 +5,12 @@ menuconfig LIBUKFALLOC
 
+config LIBUKFALLOC_STATS
+	bool
+	select LIBUKLIBID
```
Is this dependency already covered by `LIBUKSTORE`? I also do not see something directly used from the included `<uk/libid.h>` header (see below).

skuenzer commented on lib/ukfalloc/stats.c:
```diff
@@ -0,0 +1,155 @@
+#include <uk/falloc.h>
+#include <uk/falloc/store.h>
+#include <uk/init.h>
+#include <uk/libid.h>
```
Is something directly used from this header?

skuenzer commented on lib/ukfallocbuddy/fallocbuddy.c:
```diff
@@ -641,6 +646,11 @@ static inline void bfa_fl_add_tail(struct buddy_framealloc *bfa,
 
 	UK_ASSERT(mb->level < BFA_LEVELS);
 	bfa->fa.free_memory += BFA_Lx_SIZE(mb->level);
+#ifdef CONFIG_LIBUKFALLOC_STATS
```
Hum, I would prefer if we aim to have the instrumentation very small in actual implementations. What do you think of providing instrumentation macros like we did in `ukalloc`?: https://github.com/unikraft/unikraft/blob/0562d8f1c9857619586fb7901e2c54c056121875/lib/ukalloc/include/uk/alloc_impl.h#L161

michpappas replied:
Sure, will do 👍🏼

skuenzer commented on lib/ukfallocbuddy/fallocbuddy.c:
```diff
@@ -1492,7 +1519,11 @@ int uk_fallocbuddy_init(struct uk_falloc *fa)
 
-	return 0;
+#ifdef CONFIG_LIBUKFALLOC_STATS
+	rc = uk_falloc_init_stats(fa);
```
Similar thought for doing the initialization with a macro like this:
https://github.com/unikraft/unikraft/blob/0562d8f1c9857619586fb7901e2c54c056121875/lib/ukalloc/include/uk/alloc_impl.h#L228

michpappas commented:
Leaving a note here on additional work required for atomic access of per-allocator counts: https://github.com/unikraft/unikraft/issues/1548

skuenzer reviewed (approved):
Thanks a lot and I agree to address the atomic property of total_mem and free_mem with a separate PR since it crosses other systems as you describe in #1548 .

Reviewed-by: Simon Kuenzer <simon@unikraft.io>
Approved-by: Simon Kuenzer <simon@unikraft.io>
